### PR TITLE
Service::SM: Wait till client is registered

### PR DESCRIPTION
**Pull request by B3n30**
This wasn't necessary for HLE services but the run multiple LLE services it is.
There is no timeout for the wait, similar to a real 3ds

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4069)

### CODE REVIEW

File: src/core/hle/service/sm/srv.cpp [hunk 1]
```diff
@@ -99,12 +100,47 @@ void SRV::GetServiceHandle(Kernel::HLERequestContext& ctx) {
...
+            rb.Push(session.Code());
+            rb.PushMoveObjects(std::move(session).Unwrap());
+        } else if (session.Code() == Kernel::ERR_MAX_CONNECTIONS_REACHED && wait_until_available) {
+            LOG_WARNING(Service_SRV, "called service={} -> ERR_MAX_CONNECTIONS_REACHED", name);
+            // TODO(Subv): Put the caller guest thread to sleep until this port becomes available
+            // again.
+            UNIMPLEMENTED_MSG("Unimplemented wait until port {} is available.", name);
+        } else {
+            LOG_ERROR(Service_SRV, "called service={} -> error 0x{:08X}", name, session.Code().raw);
+            IPC::RequestBuilder rb(ctx, 0x5, 1, 0);
```
Comment (NarcolepticK) [inline src/core/hle/service/sm/srv.cpp]: Should this be IPC::RequestBuilder rb(ctx, 0x5, 1, 2);? Maybe form the RequestBuilder once for all of this, unless the return can vary.

File: src/core/hle/service/sm/srv.cpp [hunk 2]
```diff
@@ -99,12 +100,47 @@ void SRV::GetServiceHandle(Kernel::HLERequestContext& ctx) {
...
-        return;
+        if (wait_until_available) {
+            LOG_ERROR(Service_SRV, "called service={} delayed", name);
+            Kernel::SharedPtr<Kernel::Event> get_service_handle_event =
+                ctx.SleepClientThread(Kernel::GetCurrentThread(), "GetServiceHandle",
+                                      std::chrono::nanoseconds(-1), get_handle);
+            get_service_handle_delayed_map[name] = std::move(get_service_handle_event);
+            return;
+        } else {
+            IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
```
Comment (NarcolepticK) [inline src/core/hle/service/sm/srv.cpp]: rp.MakeBuilder(1, 2)?

File: src/core/hle/service/sm/srv.cpp [hunk 3]
```diff
@@ -99,12 +100,47 @@ void SRV::GetServiceHandle(Kernel::HLERequestContext& ctx) {
...
-        rb.Push(client_port.Code());
-        LOG_ERROR(Service_SRV, "called service={} -> error 0x{:08X}", name, client_port.Code().raw);
-        return;
+        if (wait_until_available) {
+            LOG_ERROR(Service_SRV, "called service={} delayed", name);
+            Kernel::SharedPtr<Kernel::Event> get_service_handle_event =
+                ctx.SleepClientThread(Kernel::GetCurrentThread(), "GetServiceHandle",
+                                      std::chrono::nanoseconds(-1), get_handle);
+            get_service_handle_delayed_map[name] = std::move(get_service_handle_event);
+            return;
```
Comment (NarcolepticK) [inline src/core/hle/service/sm/srv.cpp]: Missing the RequestBuilder and pushes?

File: src/core/hle/service/sm/srv.cpp [hunk 4]
```diff
@@ -99,12 +100,47 @@ void SRV::GetServiceHandle(Kernel::HLERequestContext& ctx) {
...
+        }
+    };
+
     auto client_port = service_manager->GetServicePort(name);
     if (client_port.Failed()) {
-        IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
-        rb.Push(client_port.Code());
-        LOG_ERROR(Service_SRV, "called service={} -> error 0x{:08X}", name, client_port.Code().raw);
-        return;
+        if (wait_until_available) {
```
Comment (wwylele) [inline src/core/hle/service/sm/srv.cpp]: Shouldn't this only wait on special case (error code) including not-registered and connection-full? You don't want to delay it if it errors due to permission check.

File: src/core/hle/service/sm/srv.cpp [hunk 5]
```diff
@@ -99,12 +100,47 @@ void SRV::GetServiceHandle(Kernel::HLERequestContext& ctx) {
...
+        auto client_port = service_manager->GetServicePort(name);
+
+        auto session = client_port.Unwrap()->Connect();
+        if (session.Succeeded()) {
+            LOG_DEBUG(Service_SRV, "called service={} -> session={}", name,
+                      (*session)->GetObjectId());
+            IPC::RequestBuilder rb(ctx, 0x5, 1, 2);
+            rb.Push(session.Code());
+            rb.PushMoveObjects(std::move(session).Unwrap());
+        } else if (session.Code() == Kernel::ERR_MAX_CONNECTIONS_REACHED && wait_until_available) {
```
Comment (wwylele) [inline src/core/hle/service/sm/srv.cpp]: This should be simply UNREACHABLE as srv only wakes up to call this callback on available and not-maxed port?

**Pull request merged**